[wgsl] Elaborate storage classes section

dneto0 opened the pull request:
Also:
- Define IO-shareable
- Define read-only texture
- Define write-only texture
- Add more cross-links.
- Restrict where runtime-sized arrays may be used
- Store type of uniform and storage_buffer variables must be
block-decorated

Fixes #1004

kvark commented on wgsl/index.bs:
```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+
 TODO: *This is a stub*: Collectively, the *stride* and *offset* attributes are called *layout attributes*.
 
 ### Storage Classes TODO ### {#storage-class}
```
is this still TODO?

dneto0 replied:
The description of layout is for a future MR.  So the TODO is still warranted.

kvark commented on wgsl/index.bs:
```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
 * [[#array-types]] if the array type has a stride attribute and its element type is host-shareable
 * [[#struct-types]] if all its members are host-shareable
```
This list is essentially the same  as IO-shareable, but with the addition of runtime-sized arrays?
Perhaps, it's more clear to specify the difference directly?

dneto0 replied:
I'll add a note highlighting the difference.

kvark commented on wgsl/index.bs:
```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+      <td>[=Module scope=]
+      <td>[=Host-shareable=]
+      <td>See [[#memory-layout]].
+  <tr><td>image
```
Are we using "texture" and "image" interchangeably here? I understand that the term came from SPIR-V, where the relevant Vulkan API objects are images. In WGSL, it would be great to be consistent with WebGPU.

dneto0 replied:
I'll change to "texture"

kvark commented on wgsl/index.bs:
```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+      <td>Invocations in the same [=shader stage=]
+      <td>[=Module scope=]
+      <td>[=Host-shareable=]
+      <td>See [[#memory-layout]].
```
I find it confusing that we have 3 different uniform-ish storage classes. Was hoping to navigate to that `[[#memory-layout]` to understand more, but it's empty :)

dneto0 replied:
Right, the memory-layout section is empty right now. That is yet another big change I would like to do separately.

There are 3 uniform-ish things:
- storage_buffer.  For OpenGL's SSBOs. Read-write buffers with explicit layout. Contents shareable with the host. Pretty flexible layout rules 
      - What's confusing is that originally Vulkan/SPIR-V had this in "Uniform" storage class with "BufferBlock" decorated struct.  But now the style is "StorageBuffer" with Block-decorated struct
- uniform. For OpenGL's UBOs.  Read-only buffer with explicit layout. Strict and annoying array layouts. Limit max size. Contents shareable with host.
- uniform_constant.  For opaque things that happen to be handles, each with their own binding. The handles are immutable, but the underlying data might not be.    "UniformConstant" is not my favourite name in the world but I suggest we bikeshed that separately.

kvark replied:
This separation is rather confusing. First of all, it doesn't include storage images. I [tried to check](http://shader-playground.timjones.io/427de82dab2d4ae80fba7a792f018a6d) which storage class does glslang assign to them, and apparently I'm getting "UniformConstant" for a write-only storage image... It makes so little sense to me.

I guess the bigger question to ask is - why does WGSL need storage classes *at all*?

dneto0 replied:
For one thing, you need to separate workgroup storage from private, for one thing.

Also, you have different rules for different storage classes. Layout rules differ.  

Fundamentally, storage classes are a way of organizing the many rules about how memory treatedl

kvark reviewed (commented):
Thank you for describing this!

dj2 commented on wgsl/index.bs:
```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+      <td>Same invocation only
+      <td>[=Module scope=]
+      <td>[=IO-shareable=]
+      <td>Contains values written by an upstream pipeline stage, or by the implementation.
```
Maybe

```
Input from an upstream pipeline stage, or the implementation
```

Trying to remove the word `written` as it's confusing when read with the output line below.

dneto0 replied:
I didn't want it to read as circular. But your wording looks good.

dj2 commented on wgsl/index.bs:
```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+      <td>Same invocation only
+      <td>[=Module scope=]
+      <td>[=IO-shareable=]
+      <td>Shader writes values for downstream pipeline processing.
```
Maybe

```
Output for a downstream pipeline stage. Zero initialized if required.
```

dneto0 replied:
I'll take the first part. 

The zero-initialized aspect is already described in "Variable and const" https://gpuweb.github.io/gpuweb/wgsl.html#variable-declaration

dj2 commented on wgsl/index.bs:
```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+      <td>Same invocation only
+      <td>[=Function scope=]
+      <td>[=Storable=]
+      <td>
```
```
Zero initialized if required.
```

dneto0 replied:
Covered elsewhere.

dj2 commented on wgsl/index.bs:
```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+      <td>Same invocation only
+      <td>[=Module scope=]
+      <td>[=Storable=]
+      <td>
```
```
Zero initialized if required.
```

dneto0 replied:
Covered elsewhere

dj2 commented on wgsl/index.bs:
```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+      <td>See [[#memory-layout]]
+  <tr><td>storage_buffer
+      <td>Readable.<br>
+          Also writable if the variable is not read-only.
```
By not read-only, you mean not `const`? Or is this the future NonWritable decoration?

dneto0 replied:
This is the future NonWritable.  I've added a TODO that this part of the text may change depending on the outcome of #935

dj2 commented on wgsl/index.bs:
```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+      <td>Read-only
+      <td>Invocations in the same shader stage
+      <td>[=Module scope=]
+      <td>Opaque representation of handle to a sampler or image
```
So, what is this one for? Passing samplers and images from the host side? Does a `var<uniform_constant>` have to be a `comparison_sampler`, `sampler` or one of the texture types?

dneto0 replied:
For the feature set in WebGPU, yes.  (For example, in Vulkan this also covers raytracing acceleration structures.)

The key aspect is that the thing itself has an opaque representation to the shader and to the host code.

dneto0 replied:
Also that the object itself is not mutable.  (But the thing it references might be mutable, e.g. for a storage texture)

dj2 commented on wgsl/index.bs:
```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+      <td>
+</table>
+
+The [=store type=] of a variable for the `storage_buffer` or `uniform` storage classes
```
Would this sentence make more sense to put up in the table with the storage classes? Seems like it might be missed down here.

dneto0 replied:
I've moved it up.  It's a bit crowded, but harder to miss.

dj2 commented on wgsl/index.bs:
```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+The last member of the structure type defining the store type for a `storage_buffer` variable
+may be a runtime-sized array.
+A runtime-sized array must not be used as the store type or contained within
+a store type in any other cases.
```
`[= store type=]` ?

dneto0 replied:
Done. Also, I moved this up to the section on array types.  That section still needs a bunch of other text as well.

dj2 commented on wgsl/index.bs:
```diff
@@ -432,7 +530,7 @@ storage_class
 TODO: *The following is a stub*
 
-Variables in certain storage classes must have host-shareable store type with fully elaborated memory layout.
+Variables in certain storage classes must have [=host-shareable=] [=store type=] with fully elaborated memory layout.
```
nit: `must have a `

dj2 commented on wgsl/index.bs:
```diff
@@ -443,7 +541,7 @@ This affects evaluation of a variable in one of those storage classes, or a poin
   </thead>
   <tr><td>ptr<*SC*,*T*><td>Pointer (or reference) to storage in [[#storage-class]] *SC*
-                            which can hold a value of the [[#storable-types]] *T*.
+                            which can hold a value of the [=storable=]] *T*.
```
nit: A '[' was removed but the trailing ']' remains

dj2 commented on wgsl/index.bs:
```diff
@@ -558,6 +656,11 @@ type.
 
 ### Read-only Storage Texture Types ### {#texture-ro}
+
+A <dfn noexport>read-only texture</dfn> is an image from which individual texels
```
Should this `image` be `texture`? `A read-only texture is a texture from which ...` The switch from texture to image seems confusing

dj2 commented on wgsl/index.bs:
```diff
@@ -577,6 +680,11 @@ type.
 
 ### Write-only Storage Texture Types ### {#texture-wo}
+
+A <dfn noexport>write-only texture</dfn> is an image to which individual texels
```
ditto

dneto0 commented:
I think I've addressed all feedback.  
I created a gist preview https://gistpreview.github.io/?2ab0184ed17ad6e4caea0a4f64af5794#storage-class

dj2 commented on wgsl/index.bs:
```diff
@@ -395,34 +418,127 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
   | WORKGROUP
   | UNIFORM_CONSTANT
   | STORAGE_BUFFER
   | IMAGE
```
You renamed this to `texture` in the table below, should this also change to match?

dneto0 replied:
It turns out we would only need this to express a pointer-to-image.
But that's only needed if we do image atomics.

However, we've learned that WGSL won't support image atomics. (see https://github.com/gpuweb/gpuweb/issues/728#issuecomment-696896462 ).

So the best thing is to remove "image" storage class altogether.

dj2 commented on wgsl/index.bs:
```diff
@@ -395,34 +418,127 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+      <td>Read-only
+      <td>Invocations in the same shader stage
+      <td>[=Module scope=]
+      <td>Opaque representation of handle to a sampler or texture
```
What about if we dropped uniform_constant and we just used constant. Then, on the SPIR-V side if the variable is sampler or texture type we make a UniformConstant buffer?

dneto0 replied:
I put an alternate proposal in: https://github.com/gpuweb/gpuweb/issues/1105#issuecomment-702844581

dneto0 commented:
I've removed the "image" (texture) storage class, and updated the gist:  https://gistpreview.github.io/?2ab0184ed17ad6e4caea0a4f64af5794#storage-class 

I think the main thing to focus on is what storage classes we need for textures, samplers, and buffers. #1105

dj2 commented on wgsl/index.bs:
```diff
@@ -2943,26 +3111,22 @@ Issue: (dneto) Default rounding mode is an implementation choice.  Is that what
   <tr><td>`FOR`<td>for
   <tr><td>`FRAGMENT`<td>fragment
   <tr><td>`FUNCTION`<td>function
+  <tr><td>`HANDLE`<td>handle
```
Is this needed?

dneto0 replied:
I moved it to reserved words.

If we use the full planned generality, then you can always get the base pointer of a variable and stuff it into a const.  But we could make a special rule for that. I added a TODO for that in the PR.

dneto0 commented:
I've rebased, and incorporated the feedback from this week's meeting.
I've updated the gist, and added more explanatory text in various places.

One choice that we can bikeshed about, is the name of the storage classes for UBO and SSBO.
I went with `uniform` and `buffer` because:

- that matches GLSL and is short
- when I tried "constant" I saw a lot of potential for confusion because "constant" is used all over the place, especially because `const` is a first class concept.

Note that the WebGPU spec has `uniform-buffer` and `storage-buffer`which are explicit but longer than I'd like.

dneto0 commented:
Gist URL:   https://gistpreview.github.io/?2ab0184ed17ad6e4caea0a4f64af5794#storage-class

PTAL

dneto0 commented:
Rebased and resolved a conflict.

kvark commented:
For consistency, let's use "Uniform" and "Storage" class names instead of "Uniform" and "Buffer".

litherum commented:
Discussed 2020-10-13 https://docs.google.com/document/d/16vNf_rmJ566SoI-iOGGr-UqTGZ32SHKW2_qizoB4HxU/edit#heading=h.t2r4soa0e5hc

dneto0 commented:
I've rebased against main, and updated the storage class name "buffer"->"storage".

I've updated the gist as well:  https://gistpreview.github.io/?2ab0184ed17ad6e4caea0a4f64af5794#storage-class
